granular_flow.go

```diff
@@ -41,227 +39,257 @@ func (c *baseConn) granularFlow(
-
-	buf.BeginMessage(message.Sync)
-	buf.EndMessage()
-
-	err = c.writeAndRead(ctx, buf.Unwrap())
-	if err != nil {
-		return ids, err
+func (c *baseConn) prepare(r *buff.Reader, q query) (idPair, error) {
+	c.writer.BeginMessage(message.Prepare)
+	c.writer.PushUint16(0) // no headers
+	c.writer.PushUint8(q.fmt)
+	c.writer.PushUint8(q.expCard)
+	c.writer.PushBytes([]byte{}) // no statement name
+	c.writer.PushString(q.cmd)
+	c.writer.EndMessage()
+
+	c.writer.BeginMessage(message.Sync)
+	c.writer.EndMessage()
+
+	if e := c.writer.Send(c.conn); e != nil {
+		return idPair{}, e
 	}
 
-	for buf.Next() {
-		switch buf.MsgType {
+	var (
+		err error
+		ids idPair
+	)
+
+	done := buff.NewSignal()
+
+	for r.Next(done.Chan) {
+		switch r.MsgType {
 		case message.PrepareComplete:
-			buf.PopUint16() // number of headers, assume 0
+			r.Discard(2) // number of headers, assume 0
 
 			// todo assert cardinality matches query
-			buf.PopUint8() // cardianlity
+			_ = r.PopUint8() // cardianlity
 
-			ids = idPair{
-				in:  buf.PopUUID(),
-				out: buf.PopUUID(),
-			}
+			ids = idPair{in: r.PopUUID(), out: r.PopUUID()}
 		case message.ReadyForCommand:
-			buf.PopUint16() // header count (assume 0)
-			buf.PopUint8()  // transaction state
+			// header count (assume 0)
+			// transaction state
+			r.Discard(3)
+
+			done.Signal()
 		case message.ErrorResponse:
-			return ids, decodeError(buf)
+			err = wrapAll(err, decodeError(r))
+			done.Signal()
 		default:
-			return ids, fmt.Errorf(
-				"unexpected message type: 0x%x",
-				buf.MsgType,
-			)
+			if e := c.fallThrough(r); e != nil {
+				// the connection will not be usable after this x_x
+				return idPair{}, e
+			}
 		}
 	}
 
-	return ids, nil
+	if r.Err != nil {
+		return idPair{}, r.Err
+	}
+
+	return ids, err
 }
 
-func (c *baseConn) describe(ctx context.Context) (descPair, error) {
-	buf := buff.New(c.buffer[:0])
-	buf.BeginMessage(message.DescribeStatement)
-	buf.PushUint16(0) // no headers
-	buf.PushUint8(aspect.DataDescription)
-	buf.PushUint32(0) // no statement name
-	buf.EndMessage()
+func (c *baseConn) describe(r *buff.Reader) (descPair, error) {
+	c.writer.BeginMessage(message.DescribeStatement)
+	c.writer.PushUint16(0) // no headers
+	c.writer.PushUint8(aspect.DataDescription)
+	c.writer.PushUint32(0) // no statement name
+	c.writer.EndMessage()
 
-	buf.BeginMessage(message.Sync)
-	buf.EndMessage()
+	c.writer.BeginMessage(message.Sync)
+	c.writer.EndMessage()
 
 	var descs descPair
-	err := c.writeAndRead(ctx, buf.Unwrap())
-	if err != nil {
-		return descs, err
+	if e := c.writer.Send(c.conn); e != nil {
+		return descPair{}, e
 	}
 
-	for buf.Next() {
-		switch buf.MsgType {
+	var err error
+	done := buff.NewSignal()
+
+	for r.Next(done.Chan) {
+		switch r.MsgType {
 		case message.CommandDataDescription:
-			buf.PopUint16() // num headers is always 0
-			buf.PopUint8()  // cardianlity
+			// num headers is always 0
+			// cardianlity
+			// input descriptor ID
+			r.Discard(19)
```

Comment:
Why are you discarding descriptor IDs?

Reply:
Because I already have them in scope. Perhaps there is something I'm over looking here though.

Reply:
> Because I already have them in scope. 

I don't follow what you meant here, please elaborate.

Reply:
Sorry it's not in scope in this function. The PrepareComplete response returns the descriptor IDs too. Those values are used higher up in the call stack with the descriptor values obtained here. Is this the wrong thing to do?